pkg/spanconfig/spanconfigsqlwatcher/sql_watcher.go

```diff
@@ -0,0 +1,453 @@
+	}
+	// There's nothing to do here if the database has been dropped. If the
+	// database was non-empty those objects will have their own rangefeed events.
+	if desc.Dropped() {
+		return nil, nil
+	}
+	descriptors, err := descsCol.GetAllDescriptorsInDatabase(ctx, txn, id)
+	if err != nil {
+		return nil, err
+	}
+	ret := make(descpb.IDs, 0, len(descriptors))
+	for _, desc := range descriptors {
+		ret = append(ret, desc.GetID())
+	}
+	return ret, nil
+}
+
+// onDescIDUpdate generates the span configurations for the given descriptor ID
+// and sends them to the provided updatesCh.
+func (s *SQLWatcher) onDescIDUpdate(
+	ctx context.Context, descID descpb.ID, updatesCh chan spanconfig.Update,
+) error {
+	var updates []spanconfig.Update
+	if err := descs.Txn(
+		ctx,
+		s.settings,
+		s.leaseManager,
+		s.ie,
+		s.DB,
+		func(ctx context.Context, txn *kv.Txn, descsCol *descs.Collection) error {
+			affectedIDs, err := getAllAffectedIDs(ctx, descID, txn, descsCol)
+			if err != nil {
+				return err
+			}
+			entries, err := s.generateSpanConfigurations(ctx, txn, affectedIDs)
+			if err != nil {
+				return err
+			}
+			for _, entry := range entries {
+				update := spanconfig.Update{
+					Entry:   entry,
+					Deleted: false,
+				}
+				{
+					// Try to get the table descriptor regardless of DROP status. The span
+					// config entry only needs to be removed when the descriptor is
+					// deleted, not when it is dropped, so we set the Deleted flag only if
+					// no descriptor with the given ID exists.
+					_, err := descsCol.GetImmutableDescriptorByID(ctx, txn, descID, tree.CommonLookupFlags{
+						AvoidCached:    true,
+						IncludeDropped: true,
+					})
+					if err != nil {
+						if errors.Is(err, catalog.ErrDescriptorNotFound) {
+							update.Deleted = true
+						} else {
+							return err
+						}
+					}
+				}
+				updates = append(updates, update)
+			}
+			return nil
+		}); err != nil {
+		return err
+	}
+
+	// Send all the entries on the updatesCh.
+	for _, update := range updates {
+		select {
+		case <-s.stopper.ShouldQuiesce():
+		case <-ctx.Done():
+		case updatesCh <- update:
+		}
+	}
+	return nil
+}
+
+// generateSpanConfigurations generates the span configurations corresponding to
+// the provided list of IDs. It uses a transactional view of system.zones and
+// system.descriptors to do so.
+func (s *SQLWatcher) generateSpanConfigurations(
+	ctx context.Context, txn *kv.Txn, ids descpb.IDs,
+) ([]roachpb.SpanConfigEntry, error) {
+	copyKey := func(k roachpb.Key) roachpb.Key {
+		k2 := make([]byte, len(k))
+		copy(k2, k)
+		return k2
+	}
+
+	ret := make([]roachpb.SpanConfigEntry, 0, len(ids))
+	for _, id := range ids {
+		zone, err := sql.GetHydratedZoneConfigForTable(ctx, txn, s.codec, id)
+		if err != nil {
+			return nil, err
+		}
+		spanConfig, err := zone.ToSpanConfig()
+		if err != nil {
+			return nil, err
+		}
+
+		tablePrefix := s.codec.TablePrefix(uint32(id))
+		prev := tablePrefix
```

Comment:
prevEndKey? Better yet, store the previous span. Then the check becomes !prev.EndKey.Equal(span.Key). If prev is of type roachpb.Span, do we need the defensive copyKey?